kernel/src/history_manager.rs

```diff
@@ -0,0 +1,160 @@
+use std::{
+    cmp::Ordering,
+    sync::{Arc, LazyLock},
+};
+
+use url::Url;
+
+use crate::{
+    actions::COMMIT_INFO_NAME,
+    expressions::column_name,
+    log_segment::LogSegment,
+    path::ParsedLogPath,
+    schema::{ColumnNamesAndTypes, DataType, StructField, StructType},
+    utils::require,
+    DeltaResult, Engine, Error, RowVisitor, Version,
+};
+
+pub struct LogTimeConverter {
+    timestamp: i64,
+    table_root: Url,
+}
+
+pub(crate) struct RangeResult {
+    pub start_version: Version,
+    pub start_timestamp: i64,
+}
+pub enum Bound {
+    LeastUpper,
+    GreatestLower,
+}
+impl LogTimeConverter {
+    pub(crate) fn new(table_root: Url, timestamp: i64) -> Self {
+        Self {
+            table_root,
+            timestamp,
+        }
+    }
+
+    pub(crate) fn new_range(table_root: Url, start: i64, end: i64) -> Self {
+        Self {
+            table_root,
+            timestamp: start,
+        }
+    }
+
+    fn commit_file_to_timestamp(
+        engine: &dyn Engine,
+        commit_file: &ParsedLogPath,
+    ) -> DeltaResult<(Version, i64)> {
+        let ict_type = StructField::new("inCommitTimestamp", DataType::LONG, true);
+        let schema = Arc::new(StructType::new(vec![StructField::new(
+            COMMIT_INFO_NAME,
+            StructType::new([ict_type]),
+            true,
+        )]));
+        let mut commit_timestamp = commit_file.location.last_modified;
+        let action_iter = engine.json_handler().read_json_files(
+            &[commit_file.location.clone()],
+            schema.clone(),
+            None, // not safe to apply data skipping yet
+        )?;
+        if let Some(batch) = action_iter
+            .take_while(|res| res.as_ref().is_ok_and(|batch| batch.len() > 0))
+            .next()
+        {
+            let batch = batch?;
+            let mut visitor = TimestampVisitor {
+                commit_timestamp: &mut commit_timestamp,
+            };
+            visitor.visit_rows_of(batch.as_ref())?;
+        }
+        Ok((commit_file.version, commit_timestamp))
+    }
+
+    pub(crate) fn convert(
+        engine: &dyn Engine,
+        table_root: Url,
+        timestamp: i64,
+        bound: Bound,
+    ) -> DeltaResult<Version> {
```

Comment:
@zachschuermann  there's probably some more optimization we can do here for range queries. A basic one is using the idx of start_version as `lo` instead of starting `lo=0`.

Another is to remember more from the search for start, and you might be able to get a tighter bound for `lo`.  Given that each IO is expensive, this is probably a good optimization.